### ISSUE TITLE
Electron defaults to en-US and MM/dd/YY date formats regardless of system settings

### ISSUE THREAD

**Issue by craxal** [electron/electron#13023]
* Electron Version: 2.0.0, 4.2.5
* Operating System (Platform and Version): Windows 10
* Last known working Electron version: Unknown

**Expected Behavior**
Electron should default to the OS's current culture/locale and date format settings.

**Actual behavior**
Electron defaults to en-US and the MM/dd/YY format for dates. More specifically:
- Entering `navigator.language` should return whatever my OS is set to.
- Entering `new Date().toLocaleString()` should return a date string in the format set by my OS.

**To Reproduce**
1. If in the US, change your OS language/region to something else (like United Kingdom or Sweden).
2. Install and run the Electron Quick Start sample application.
3. Open the developer tools and enter `navigator.language` in the console window.
4. In the console window, enter `new Date().toLocaleString()`.

**Comment by aabuziad** [comment]
**Workaround: We can create and call C# function to retrieve system date format, using 'edge' package that allow us connect between Node.js and .NET**

**Steps:**
- npm install electron-edge-js(https://www.npmjs.com/package/electron-edge-js)
- create native-functions.js file next to Electron 'main.ts', inside the created file add the following code:  
```
var edge = require('electron-edge-js');
var getSystemDateFormat = edge.func(function () {/*
    using System.Globalization;
    async (param) => {
        return CultureInfo.CurrentCulture.DateTimeFormat.ShortDatePattern;
    }
*/});
module.exports = getSystemDateFormat;
```

- inside main.ts add the following code:
```
import * as getSystemDateFormat from './native-functions.js';
getSystemDateFormat('some data', function (error, result) {
  if (error) throw error;
  console.log(result);
});
```

**Steps explanation:**
- First step we installed the required dependencies
- Second step we created C# native function
- Third step we called the C# function
**system date format is stored in result and will be printed to console**

**More:**
- electron-edge-js is a clone of edge(https://www.npmjs.com/package/edge), its made for electron usage and it's easier to setup because it installs the required dependencies and compatible with different Electron and Node.js versions.
- electron-edge-js also comes precompiled with correct Node.js versions
- if you change native-functions.js to native-functions.ts, the commented C# code may be **omitted** after compiling typescript(.ts) to javascript(.js)
- you can send parameters to C# code, in the above example we've sent a string value **'some data'**  to C# code, this string value is stored in **param**

**Tested On:**
- Windows 10 (x64)
- Node.js v8.11.4
- Electron 2.0.2
- electron-edge-js 8.3.8

**Comment by ckerr** [comment]
This may be a Windows-only issue -- I'm not able to reproduce this on Linux using either Electron 2.0.0 or 3.0.2 to run electron-quick-start:

```sh
$ LANG=de_DE npm start
```
And then, in devtools:

```text
> navigator.language
< "de"
> new Date().toLocaleString()
< "28.9.2018, 14:37:14"
```

**Comment by craxal** [comment]
To offer some clarification (since I was a bit confused myself), there are two concepts at work here:

#### The local machine's locale (region, language)

Getting the machine's locale is already solved: `app.getLocale()`. This works as expected.

#### The local machine's regional formatting (dates, times)

Getting the machine's regional formatting is the problem here. If the machine's regional formatting is left unaltered from the default for it's current locale, there's no problem. But lots of people customize their regional settings. For example, the default date format in the US is MM/DD/YYYY. Some of us prefer DD/MM/YYYY, so we change the regional formatting settings. Electron has no API for retrieving system regional formats, neither does it seem to respect them by default.

**Comment by lalomartins** [comment]
It's a bit more than that. Electron's formatting takes only the language in consideration. If my region is set to Estonia or Germany or Japan and language is en-US, it will format MM/DD/YYYY. The *proper* thing to do would be to read my formatting setting (YYYY-MM-DD), but if it actually used the region default (DD/MM/YYYY), the problem would be less serious.

**Comment by jefgen** [comment]
To be honest, I don't think this problem is specific to Electron -- this is more of a limitation of Chromium itself.
For example, there are a number of bug reports on Chromium that go back for years about this issue:

https://bugs.chromium.org/p/chromium/issues/detail?id=234897
https://bugs.chromium.org/p/chromium/issues/detail?id=94906

Unfortunately, they don't seem to get much traction though.

**Comment by sofianguy** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. 

If you're still experiencing this issue in Electron v4.2.x or later, please add a comment specifying the version you're testing with and any other new information that a maintainer trying to reproduce the issue should know.

I'm setting the `blocked/need-info` label for the above reasons. This issue will be closed 7 days from now if there is no response.

Thanks in advance! Your help is appreciated.

**Comment by lalomartins** [comment]
Still present on 4.2.9:
![image](https://user-images.githubusercontent.com/1211455/62751744-8198ff00-ba6d-11e9-9460-72627b564be1.png)

Still present in 5.0.9: 
![image](https://user-images.githubusercontent.com/1211455/62751856-f2d8b200-ba6d-11e9-9d16-b3c7933380fe.png)

Still present in 6.0.1: 
![image](https://user-images.githubusercontent.com/1211455/62751967-57940c80-ba6e-11e9-941a-5b63d09928fa.png)

Still present in 7.0.0-beta.2: 
![image](https://user-images.githubusercontent.com/1211455/62752067-a5107980-ba6e-11e9-9924-1ec1a8fbd4c7.png)

My region and format are set to Japan, so typing `date` on PowerShell returns `2019年8月9日 6:27:58`.

**Comment by TuckerTR** [comment]
Adding an athlete experiencing this issue:

## Athlete 1

**Username:** [Bill Evans](https://www.trainerroad.com/app/career/bill%20evans)
**Zendesk Ticket:** [195851](https://trainerroad.zendesk.com/agent/tickets/195851)
**Edit User Page:** https://www.trainerroad.com/app/admin/user/157295/edit
**User Logs:** https://www.trainerroad.com/app/admin/users/157295/logs
**Date first occured:** 1/3/2021

**Comment by craxal** [comment]
@TuckerTR Are you posting to the correct page? This issue page is for Microsoft Azure Storage Explorer, not the website you linked to.

**Comment by electron-triage** [comment]
The Electron version reported on this issue is no longer supported. See our [supported versions documentation](https://www.electronjs.org/docs/tutorial/support#supported-versions).

If this is still reproducible on a supported version, please open a new issue with any other new information that a maintainer should know.

Thank you for taking the time to report this issue and helping to make Electron better! Your help is appreciated.

**Comment by craxal** [comment]
Hi there. We're up to Electron 13 these days, and I'm wondering if there's any update about how this issue might be resolved.

**Comment by thomasdao** [comment]
I tested this code in Chrome and Electron in my Mac:

```javascript
new Intl.DateTimeFormat([], {
  dateStyle: 'medium',
  timeStyle: 'short'
}).format(new Date())
```

In Chrome `101.0.4951.54`, it follows the regional format defined in the system settings: `6 May 2022, 11:22`

In Electron `18.2.0` and `19.0.0-beta.4`, it always uses the US date format: `May 6, 2022, 11:00 AM`

It looks like Chrome is using the right format and Electron is not working correctly, even in the newer version 19.0.0-beta.4 which based on Chrome 102.

**Comment by lalomartins** [comment]
This issue is still present in the latest version.

This is a serious issue in Electron itself. I don't see what's the point of allowing it to go automatically stale, it won't ‘magically’ disappear unless someone actually writes a fix.

**Comment by lalomartins** [comment]
This issue is still present in the latest version.

This is a serious issue in Electron itself. I don't see what's the point of allowing it to go automatically stale, it won't ‘magically’ disappear unless someone actually writes a fix.

**Comment by mfalkvidd** [comment]
Here we go again

**Comment by Hedwig188** [comment]
bumpOn Jul 11, 2023, at 21:12, github-actions[bot] ***@***.***> wrote:﻿
This issue has been automatically marked as stale. If this issue is still affecting you, please leave any comment (for example, "bump"), and we'll keep it open. If you have any new additional information—in particular, if this is still reproducible in the latest version of Electron or in the beta—please include it with your comment!

**Comment by craxal** [comment]
Hey Electron folks, this has been open for quite a long time without comment from any of the devs. Can anyone please provide an update about any progress made on this issue? These days, our app is staying up-to-date on the latest Electron (we're now on v25).

**Comment by alametti** [comment]
Hi Electron people, there are open issues in many other **major** projects that are waiting for this issue to be resolved to be able to implement sane, future-proof fixes: see <https://github.com/signalapp/Signal-Desktop/issues/4252>, for example.

Any updates? This issue has been open for **over 5 years**, throughout **8 major numbered versions**, and still no fix for a basic UI function.

**Comment by john8329** [comment]
I agree this is serious, there's no way of getting the system date locale, and we have to hardcode the format.

**Comment by Kile-Asmussen** [comment]
This issue is present in VSCode 1.82.2, which uses electron. I have just spent two hours of my time chasing down this issue. This is a very sad state of i18n affairs.

**Comment by mfalkvidd** [comment]
The madness continues …

**Comment by Haraven** [comment]
I ended up here after looking up how to change date formats. 

I would very much like to see this issue addressed. I'm developing an app for a region where the default is date format is DD/MM/YYYY. It's just confusing to see MM/DD/YYYY being the default, especially for days <13 (e.g., 10/12/2023 - is it the 10th of December or the 12th of October? :shrug: ).

It's also somewhat annoying that I can't set the date format myself. I've tried [this](https://stackoverflow.com/a/21323755) and [this](https://stackoverflow.com/a/38228964), with no luck.

**Comment by gsidhu** [comment]
I am getting a weird result in my Electron app where Chromium seems to pick up the right (my default) locale –
<img width="485" alt="image" src="https://github.com/electron/electron/assets/5348728/041da8d2-304f-4f2a-8b21-15f418ae672e">

But the backend NodeJS defaults to `en-US` –
<img width="197" alt="image" src="https://github.com/electron/electron/assets/5348728/261c9123-79c3-4380-bc24-43d2b9263708">

Bumping this in the hopes of a fix from the electron contributors!

**Comment by paulgrav** [comment]
I’m a user who see this issue in every single Electron app that I encounter. I find it deeply frustrating and depressing that in 2024 we still can’t format dates as per a user’s setting.

**Comment by mfalkvidd** [comment]
.

**Comment by john8329** [comment]
Just because it's stale doesn't mean it's not relevant. This is a fundamental issue, and Electron is designed to give a better interaction with the OS

**Comment by Kile-Asmussen** [comment]
bump, this is a serious breach of i18n guidelines.

**Comment by f0def** [comment]
Bug is still actual, why you ignore?

**Comment by craxal** [comment]
@ckerr This issue has been stuck in a stale/bump cycle for several years now. It would be good to know if someone could provide any kind of update as to what whether anyone has been looking at this issue, whether it's being considered, anything at all.

To work around some of these issues, our team has implemented a native node module with functions for formatting dates and times, as well as a few other number formats, using the system's custom formats. We call these functions explicitly when displaying dates/times/numbers:
* For Windows, we use a variety of Windows functions, such as `GetLocaleInfoEx`, `GetNumberFormatEx`, `GetDynamicTimeZoneInformation`, `SystemTimeToTzSpecificLocalTimeEx`, `UnixTimeToSystemTime`, `GetDateFormatEx`, and `GetTimeFormatEx`. Custom formats are quite straightforward, but converting between JS and Windows C is complicated.
* For Mac, we use `NSDateFormatter` and `NSNumberFormatter` with various lengths. Custom formats are a little harder to set in Mac since Apple removed the UI for setting full, long, medium, and short formats.
* For Linux, we use `setLocale`, `nl_langinfo`, and `strftime`. Custom formats require creating custom locale files and registering them with the system.

**Comment by john8329** [comment]
Translated, this bug won't be fixed because it's hard. Thanks @craxal for the very useful insights.

**Comment by jkleinsc** [comment]
@craxal we rely on date formatting provided by Chromium, so if the issue exists there, we end up essentially inheriting that behaviour.  To my knowledge no one has tried to address this issue directly in Electron (but we certainly welcome contributions to address it).  Please remember that Electron as an open source project is largely volunteer-driven and collaborative. Because of this, time and energy to get things done are often the scarcest resources. We want Electron to be the best that it can be, but we also have to be realistic about what the team maintaining Electron can achieve.

Having said that, addressing this issue is complicated.  There are several approaches that could be taken:
1. Wait for Chromium to address this issue.  I've added the `blocked/upstream` label to the issue to note that currently we are blocked by upstream's implementation.
2. Add custom code/patches to Electron to address this issue.  As described in the comment above, each OS requires unique approaches to handle this issue, which most likely means a large amount of code/customization might be needed.  
3. Use a native node module to address this issue.  @craxal since this is the approach you have taken, does this approach satisfy your needs?  Are there shortcomings that a native node module can't address? Additionally, have you considered open-sourcing your node module? It sounds like it could be useful to other developers.

Additionally, I added the `stale-exempt` label to this issue to turn off the stale issue nagging since this is an issue that is persisting with no clear intention from chromium to address.

**Comment by craxal** [comment]
@jkleinsc Thank you very much for the update.

A custom native node module is sufficient for our needs. Now, instead of doing something like `new Date().toLocaleString()`, we do `formatLocalDateTime(new Date(), "short")`. We aimed for something simple to use and clear on intent.

Part of me feels like this native node module API is more of a crutch than a solution, but maybe I'm overthinking it.

* Formatting is strictly opt-in. You have to call formatting functions explicitly to format data. Then again, being explicit about what you want and when is usually a good thing.
* As of today, the module supports local/UTC dates, times, and date times of varying length, numbers to a specified precision, and bytes (B, KiB, MiB, etc.). We also support getting the system's list separator. That covers our bases. If additional formats were needed, I don't think it would be too hard to add more.
* It only works in one direction. Parsing data types from formats could be added in, but it would require a lot more work. That's not something we really need.
* Depending on the platform, formatting in date/time inputs might not match what the native node module outputs. Not that I've observed this, but as I don't know how date/time inputs are implemented, it could be possible. The same could be true for other input types.

If the wider community were interested in this API solution, I'd be happy to discuss with my team the possibility of opening up the source the code. I'm sure we'd like hearing feedback on it as well.

**Comment by joaoportela** [comment]
@jkleinsc doesn't the comment by @gsidhu indicate that chromium already addresses this? at least partially...  🤔
Or am I missing something obvious? 😅 

PS - I'm referring to [this comment](https://github.com/electron/electron/issues/13023#issuecomment-1899858812):

> I am getting a weird result in my Electron app where Chromium seems to pick up the right (my default) locale – <img alt="image" width="485" src="https://private-user-images.githubusercontent.com/5348728/297976626-041da8d2-304f-4f2a-8b21-15f418ae672e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQxNjM4OTMsIm5iZiI6MTc2NDE2MzU5MywicGF0aCI6Ii81MzQ4NzI4LzI5Nzk3NjYyNi0wNDFkYThkMi0zMDRmLTRmMmEtOGIyMS0xNWY0MThhZTY3MmUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTEyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMjZUMTMyNjMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmU2OTEwMmEzYjZhMGFjYmUxNDZiZDY0N2FkYzYxNGM3NDA3NzhhYTNkM2YzODlkMDVmNzlkNzcwNmRlYWZhZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.8zLjc0y9072Gf7PagZex7bbFMwVBCJqyAO0ywRqrsng">
> 
> But the backend NodeJS defaults to `en-US` – <img alt="image" width="197" src="https://private-user-images.githubusercontent.com/5348728/297977083-261c9123-79c3-4380-bc24-43d2b9263708.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQxNjM4OTMsIm5iZiI6MTc2NDE2MzU5MywicGF0aCI6Ii81MzQ4NzI4LzI5Nzk3NzA4My0yNjFjOTEyMy03OWMzLTQzODAtYmMyNC00M2QyYjkyNjM3MDgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTEyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMjZUMTMyNjMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2U0NjdiODgzMDY5OTZlMzg1MWE4NjcxMmU5MjZiNDRhZDY0ZjIxYjgyZWUwMDNiYTcyNGIwOTVjNTcwYzhmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.V0seol2ZXjHcKq5ZJ7lt7Wn2kW2qDf3EN3Xv_L2lG0Q">
> 
> Bumping this in the hopes of a fix from the electron contributors!
> 
> I am running –
> 
> ```
> node: '16.17.1',
> electron: '22.3.27',
> chrome: '108.0.5359.215'
> ```
